src/meta/src/barrier/mod.rs

```diff
@@ -552,27 +553,32 @@ where
             self.in_flight_barrier_nums,
         );
 
-        let mut state = BarrierManagerState::create(self.env.meta_store()).await;
-        if self.enable_recovery {
-            // handle init, here we simply trigger a recovery process to achieve the consistency. We
-            // may need to avoid this when we have more state persisted in meta store.
-
-            // XXX(bugen): why we need this?
-            let new_epoch = state.in_flight_prev_epoch().next();
-
-            self.set_status(BarrierManagerStatus::Recovering).await;
-            let span = tracing::info_span!("bootstrap_recovery", new_epoch = new_epoch.value().0);
-            let new_epoch = self.recovery(new_epoch).instrument(span).await;
-            state
-                .update_inflight_prev_epoch(self.env.meta_store(), new_epoch)
-                .await
-                .unwrap();
-        } else if self.fragment_manager.has_any_table_fragments().await {
+        if !self.enable_recovery && self.fragment_manager.has_any_table_fragments().await {
             panic!(
                 "Some streaming jobs already exist in meta, please start with recovery enabled \
-            or clean up the metadata using `./risedev clean-data`"
+                or clean up the metadata using `./risedev clean-data`"
             );
         }
+
+        let mut state = {
+            let latest_snapshot = self.hummock_manager.latest_snapshot();
+            assert_eq!(
+                latest_snapshot.committed_epoch, latest_snapshot.current_epoch,
+                "persisted snapshot must be from a checkpoint barrier"
+            );
+            let prev_epoch = TracedEpoch::new(latest_snapshot.committed_epoch.into());
+
+            // Bootstrap recovery. Here we simply trigger a recovery process to achieve the
+            // consistency.
+            // Even if there's no actor to recover, we still go through the recovery process to
+            // inject the first `Initial` barrier.
+            self.set_status(BarrierManagerStatus::Recovering).await;
+            let span = tracing::info_span!("bootstrap_recovery", prev_epoch = prev_epoch.value().0);
+            let new_epoch = self.recovery(prev_epoch).instrument(span).await;
```

Comment:
cc @yezizp2012 so that we can unify the first barrier handling (i.e. `INVALID_EPOCH` previously) with the one after failure recovery.